### PR TITLE
[UNR-3201] Fix multiple dynamic subobjects being created in the same tick

### PR DESCRIPTION
**Pull request by mironec**
#### Description
Creating multiple dynamic subobjects in the same tick would break, due to them arriving in the same critical section.

#### Release note
Fixed a bug when creating multiple dynamic subobjects at the same time, when they would fail to be created on clients.

#### Tests
Soon to come in EngineNetTest as part of the DynamicComponentTests.

#### Documentation
Release note.

#### Primary reviewers
@m-samiec @samiwh

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 1]
```diff
@@ -105,11 +105,44 @@ void USpatialReceiver::LeaveCriticalSection()
 		{
 			OnEntityAddedDelegate.Broadcast(PendingAddEntity);
 		}
+		PendingAddComponents.RemoveAll([PendingAddEntity](const PendingAddComponentWrapper& Component) {return Component.EntityId == PendingAddEntity;});
 	}
 
 	for (Worker_AuthorityChangeOp& PendingAuthorityChange : PendingAuthorityChanges)
 	{
-		HandleActorAuthority(PendingAuthorityChange);
+		if (PendingAuthorityChange.authority != WORKER_AUTHORITY_AUTHORITATIVE)
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:113]: WORKER_AUTHORITY_AUTHORITY_LOSS_IMMINENT doesn't seem to be handled here anymore, is this necessary? 

Perhaps this could loop and RemoveSwap and the loop below could process all remaining ops without condition?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 2]
```diff
@@ -105,11 +105,44 @@ void USpatialReceiver::LeaveCriticalSection()
...
+		{
+			HandleActorAuthority(PendingAuthorityChange);
+		}
+	}
+
+	for (PendingAddComponentWrapper& PendingAddComponent : PendingAddComponents)
+	{
+		if (StaticComponentView->GetComponentData<UnrealMetadata>(PendingAddComponent.EntityId) == nullptr)
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:121]: Should we log in these cases? Does this mean we don't have the actor checked out?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 1 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:113]: I think the `WORKER_AUTHORITY_AUTHORITY_LOSS_IMMINENT` will fall into this processing loop, since they are not equal to `WORKER_AUTHORITY_AUTHORITATIVE`. That being said, yes, we do go over the entire list twice, while we could removeswap. Wonder which one is faster?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 3]
```diff
@@ -105,11 +105,44 @@ void USpatialReceiver::LeaveCriticalSection()
 		{
 			OnEntityAddedDelegate.Broadcast(PendingAddEntity);
 		}
+		PendingAddComponents.RemoveAll([PendingAddEntity](const PendingAddComponentWrapper& Component) {return Component.EntityId == PendingAddEntity;});
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:108]: What's the purpose of this? Are these component adds already handled in some other way?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 1 above)
Comment (improbable-danny) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:113]: Oops, yeah - misread that! Maybe ignore this then :D

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 3 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:108]: Yeah, in ReceiveActor, we go through the relevant ones and handle them (but not remove them from the array, which is what this is doing).

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 2 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:121]: I have started writing a really long explanation, and I am no longer sure whether this is correct.
The initial reasoning for this was that I wanted to avoid handling components we didn't need to handle. `HandleIndividualAddComponent` is written in a way that assumes we are dealing with dynamic components on actors. The code above therefore discards all add components which are not associated with actors (e.g. GlobalStateManager, SpatialSpawner, etc.).
This should mostly work, but also, it should not happen.
If `HandleIndividualAddComponent` finds out that the component doesn't have an associated actor channel, it prints a warning. I think I will just remove this condition and let it fall into the logging branch of `HandleIndividualAddComponent`. But this does make me uneasy.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 4]
```diff
@@ -105,11 +105,44 @@ void USpatialReceiver::LeaveCriticalSection()
...
+
+		UE_LOG(LogSpatialReceiver, Verbose,
+			TEXT("Add component inside of a critical section, outside of an add entity, being handled: component id: %d, entity id: %d"),
+			PendingAddComponent.ComponentId, PendingAddComponent.EntityId);
+		HandleIndividualAddComponent(PendingAddComponent.EntityId, PendingAddComponent.ComponentId, MoveTemp(PendingAddComponent.Data));
+	}
+
+	for (Worker_AuthorityChangeOp& PendingAuthorityChange : PendingAuthorityChanges)
+	{
+		if (PendingAuthorityChange.authority == WORKER_AUTHORITY_AUTHORITATIVE)
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:142]: By the way, the reason the AuthorityChange processing is now split according to authority is to avoid cases where we receive data while being authoritative, as that could be unintuitive to the game devs. Now we process Lose Auth -> Add Components -> Gain Auth. A common thing that happens is that on handover we get ComponentData -> Gain Auth, and in the previous flow, we would have first gained authority and then applied updates. This is more intuitive, I think, you receive data as if you were a client to get the most up-to-date state, and then gain authority.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 3 above)
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:108]: Any reason to not do this when we're iterating over them in ReceiveActor?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 4 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:142]: This is a good comment - worth including it - or a subsection of it in the file

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:142]: This is only true for components in critical sections though, isn't it possible to get authority and a component data outside a critical section?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 5]
```diff
@@ -105,11 +105,44 @@ void USpatialReceiver::LeaveCriticalSection()
...
+		{
+			continue;
+		}
+		if (ClassInfoManager->IsGeneratedQBIMarkerComponent(PendingAddComponent.ComponentId))
+		{
+			continue;
+		}
+		if (StaticComponentView->HasAuthority(PendingAddComponent.EntityId, PendingAddComponent.ComponentId))
+		{
+			continue; // Hacky, allows servers to change state if they are going to be authoritative, without us overwriting it with old data
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:131]: This would also mean we're dropping the last updates from the runtime right? We sure we want to do this. Have you seen problem otherwise?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 6]
```diff
@@ -105,11 +105,44 @@ void USpatialReceiver::LeaveCriticalSection()
...
+		}
+
+		UE_LOG(LogSpatialReceiver, Verbose,
+			TEXT("Add component inside of a critical section, outside of an add entity, being handled: component id: %d, entity id: %d"),
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:135]:
```suggestion
			TEXT("Add component inside of a critical section, outside of an add entity, being handled: component id: %d, entity id: %lld"),
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 5 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:131]: Yes, otherwise there is a problem. Seems that PlayerState in particular is susceptible to this, as I think the initial data doesn't have the owner or something like that set. Then it's changed locally, and the AddComponent from runtime back would overwrite the owner back to nullptr when it was resolved okay locally.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 3 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:108]: I did at first, but still not sure whether we want to preserve order and I suspect we do. In which case this should be faster.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 4 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:142]: You would then receive them in the correct order from the Worker SDK, and since we don't (shouldn't) queue them outside critical sections, it should be okay.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 5 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:127]: This would break any type of soft handover we might want to implement also. I'm a little concerned about this change - we should probably create a follow up ticket and track it here, as this is just as likely to sting us as it is to help.

Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:127]: Ah, yes, definitely, this is a hacky workaround for current behaviour.
I will create a ticket.

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialReceiver.h [hunk 7]
```diff
@@ -173,7 +176,7 @@ class USpatialReceiver : public UObject, public SpatialOSDispatcherInterface
 	void QueueAuthorityOpForAsyncLoad(const Worker_AuthorityChangeOp& Op);
 	void QueueComponentUpdateOpForAsyncLoad(const Worker_ComponentUpdateOp& Op);
 
-	TArray<PendingAddComponentWrapper> ExtractAddComponents(Worker_EntityId Entity);
+	TArray<PendingAddComponentWrapper> ExtractAddComponents(TArray<PendingAddComponentWrapper>* QueuedComponents, Worker_EntityId Entity);
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialReceiver.h:179]: Make this a reference instead of a pointer.

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialReceiver.h [hunk 8]
```diff
@@ -124,12 +127,12 @@ class USpatialReceiver : public UObject, public SpatialOSDispatcherInterface
 
 	void ApplyComponentDataOnActorCreation(Worker_EntityId EntityId, const Worker_ComponentData& Data, USpatialActorChannel& Channel, const FClassInfo& ActorClassInfo, TArray<ObjectPtrRefPair>& OutObjectsToResolve);
 	void ApplyComponentData(USpatialActorChannel& Channel, UObject& TargetObject, const Worker_ComponentData& Data);
-	
-	// This is called for AddComponentOps not in a critical section, which means they are not a part of the initial entity creation.
-	void HandleIndividualAddComponent(const Worker_AddComponentOp& Op);
+
+	void HandleIndividualAddComponent(Worker_EntityId EntityId, Worker_ComponentId ComponentId, TUniquePtr<SpatialGDK::DynamicComponent>&& Data);
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialReceiver.h:131]: Pass unique_ptr by value unless this function conditionally moves it (which I don't think is the case).

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 9]
```diff
@@ -1241,32 +1306,32 @@ void USpatialReceiver::ApplyComponentDataOnActorCreation(Worker_EntityId EntityI
...
 	}
 
-	const FClassInfo& Info = ClassInfoManager->GetClassInfoByComponentId(Op.data.component_id);
-	AActor* Actor = Cast<AActor>(PackageMap->GetObjectFromEntityId(Op.entity_id).Get());
+	const FClassInfo& Info = ClassInfoManager->GetClassInfoByComponentId(ComponentId);
+	AActor* Actor = Cast<AActor>(PackageMap->GetObjectFromEntityId(EntityId).Get());
 	if (Actor == nullptr)
 	{
-		UE_LOG(LogSpatialReceiver, Verbose, TEXT("Received an add component op for subobject of type %s on entity %lld but couldn't find Actor!"), *Info.Class->GetName(), Op.entity_id);
+		UE_LOG(LogSpatialReceiver, Warning, TEXT("Received an add component op for subobject of type %s on entity %lld but couldn't find Actor!"), *Info.Class->GetName(), EntityId);
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:1334]: Nit - try to make the main body of the message copy and pastable by putting all the dynamic stuff at the end.

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialReceiver.h (hunk 8 above)
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialReceiver.h:131]: Really the dynamic component should contain a unique_ptr but in that case the same advise would apply.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 10]
```diff
@@ -2563,6 +2638,19 @@ void USpatialReceiver::MoveMappedObjectToUnmapped(const FUnrealObjectRef& Ref)
...
 
+void USpatialReceiver::ClearPendingOwnerOnlyComponentsForEntity(Worker_EntityId EntityId)
+{
+	TArray<PendingAddComponentWrapper> QueuedComponents = ExtractAddComponents(&PendingOwnerOnlyComponents, EntityId);
+
+	const FDateTime Now = FDateTime::UtcNow();
+	for (const auto& Component : QueuedComponents)
+	{
+		FTimespan QueueTime = Now - Component.Timestamp;
+		UE_LOG(LogSpatialReceiver, Verbose, TEXT("Removed queued OwnerOnly component %d for entity %lld after %f seconds"),
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:2649]: Same as elsewhere - better if the dynamic part is all at the end.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 11]
```diff
@@ -105,11 +105,53 @@ void USpatialReceiver::LeaveCriticalSection()
...
+	for (PendingAddComponentWrapper& PendingAddComponent : PendingAddComponents)
+	{
+		if (ClassInfoManager->IsGeneratedQBIMarkerComponent(PendingAddComponent.ComponentId))
+		{
+			continue;
+		}
+		USpatialActorChannel* Channel = NetDriver->GetActorChannelByEntityId(PendingAddComponent.EntityId);
+		if (Channel == nullptr)
+		{
+			UE_LOG(LogSpatialReceiver, Warning, TEXT("Got an add component for an entity (%lld) that doesn't have an associated actor channel."), PendingAddComponent.EntityId);
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:133]: Discussed offline but I'd prefer to not rely on actor channel. Using information in there couples this to whatever was adding data to it and make everything generally harder to reason about.
Also discussed offline - this probably shouldn't be a warning. Either it should be an error, verbose, or not be here.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 12]
```diff
@@ -105,11 +105,53 @@ void USpatialReceiver::LeaveCriticalSection()
...
+			continue;
+		}
+		if (Channel->bCreatedEntity)
+		{
+			// Allows servers to change state if they are going to be authoritative, without us overwriting it with old data.
+			// TODO: UNR-3457 to remove this workaround.
+			continue;
+		}
+
+		UE_LOG(LogSpatialReceiver, Verbose,
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:143]: I don't think this is a bad thing to have exactly but we seem to add verbose logs in a really adhoc basis. It might be worth investing some time in working out how we want to register certain events like this happening and which ones are actually worth taking note of.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 1 above)
Comment (samiwh) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:118]: I was going to bring this up too. It's also easier to see what you are doing if you are explicit with which ops you want to process.

Are we sure you can never get an add and remove for the same entity in this list? If so then we need to preserve the ordering.

Ultimately this should all be replaced by spatial view so I don't think performance on this matters.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 13]
```diff
@@ -239,13 +281,21 @@ void USpatialReceiver::OnAddComponent(const Worker_AddComponentOp& Op)
 		return;
 	}
 
+	// Queue owner-only components until we have authority
+	if (ClassInfoManager->GetCategoryByComponentId(Op.data.component_id) == SCHEMA_OwnerOnly &&
+		!StaticComponentView->HasAuthority(Op.entity_id, SpatialConstants::CLIENT_ENDPOINT_COMPONENT_ID))
+	{
+		PendingOwnerOnlyComponents.AddUnique(PendingAddComponentWrapper(Op.entity_id, Op.data.component_id, MakeUnique<DynamicComponent>(Op.data)));
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:288]: Can these be added and removed before gaining authority?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 14]
```diff
@@ -1241,32 +1306,32 @@ void USpatialReceiver::ApplyComponentDataOnActorCreation(Worker_EntityId EntityI
...
+void USpatialReceiver::HandleIndividualAddComponent(Worker_EntityId EntityId, Worker_ComponentId ComponentId, TUniquePtr<SpatialGDK::DynamicComponent>&& Data)
 {
 	uint32 Offset = 0;
-	bool bFoundOffset = ClassInfoManager->GetOffsetByComponentId(Op.data.component_id, Offset);
+	bool bFoundOffset = ClassInfoManager->GetOffsetByComponentId(ComponentId, Offset);
 	if (!bFoundOffset)
 	{
 		UE_LOG(LogSpatialReceiver, Warning, TEXT("EntityId %lld, ComponentId %d - Could not find offset for component id "
-			"when receiving dynamic AddComponent."), Op.entity_id, Op.data.component_id);
+			"when receiving dynamic AddComponent."), EntityId, ComponentId);
```
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:1316]: Drive-by rather than new comment, but again better to keep dynamic parts of the message at the end.
E.g. "[some_message]. Entity ID: {entityId}".

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 9 above)
Comment (samiwh) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:1334]: Why is this a warning? What has gone wrong when this happens? If it's actually a problem it should be an error, if it's a transient error that isn't a major problem but users should be aware how often it's happening then it should be warning.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 12 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:143]: My general idea for verbose logs is that Unreal really doesn't like to show them by default, so they will almost NEVER be printed. The situation where I would expect to read this, is during debugging, e.g. some game studio has a replication issue, so we tell them to turn on verbose logs so we can better see what may be going wrong. Which is the reasoning behind this one. While debugging, I would like to know whether this problematic (so far) situation has occurred.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 13 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:288]: I have thought and thought some more, and I think we can fix this in a different way, so will try a small rewrite.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 15]
```diff
@@ -105,11 +105,68 @@ void USpatialReceiver::LeaveCriticalSection()
...
+	for (PendingAddComponentWrapper& PendingAddComponent : PendingAddComponents)
+	{
+		if (ClassInfoManager->IsGeneratedQBIMarkerComponent(PendingAddComponent.ComponentId))
+		{
+			continue;
+		}
+		USpatialActorChannel* Channel = NetDriver->GetActorChannelByEntityId(PendingAddComponent.EntityId);
+		if (Channel == nullptr)
+		{
+			UE_LOG(LogSpatialReceiver, Warning, TEXT("Got an add component for an entity (%lld) that doesn't have an associated actor channel."), PendingAddComponent.EntityId);
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:133]:
```suggestion
			UE_LOG(LogSpatialReceiver, Error, TEXT("Got an add component for an entity (%lld) that doesn't have an associated actor channel."), PendingAddComponent.EntityId);
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 9 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:1329]: I think it falls into the second category, as it is an error somewhere most likely, but the impact on the user should be small, as the actor is deleted anyway.
